File: src/eopx/artifact_figure.py
"""EPX-F — the artifact figure: a frozen, recomputable face for a `.eopx`.

The figure is a **cell grid**, not a picture. Each cell carries an integer
*level*; glyphs are a presentation choice layered on top. This is what lets the
same figure appear as ASCII in a terminal, as Unicode blocks in a rich console,
or engraved on a badge, while remaining *one* object with *one* digest.

Two properties make it worth freezing (see ``docs/specs/EPX-F_artifact_figure.md``):

**It is recomputable, never attestable by eye.** The grid derives from the
*pre-image* half of the signed manifest — the fields fixed before the image
exists. A verifier recomputes the grid from a `.eopx` and compares; looking at
a pretty picture proves nothing (POSITIONING).

**It survives key rotation.** The grid has two bands. The **content band**
(rows 0..5) depends on ``merkle_root`` alone, so an artifact keeps its face for
life. The **epoch band** (rows 6..7) also takes ``dilithium_pk_fp``, so when the
issuing key rotates the artifact visibly moves to a new epoch while staying
recognisably itself. A badge minted under epoch N does not become a stranger
under epoch N+1.

Why not ``payload_hash`` or ``image_sha3_512``
----------------------------------------------
Both are downstream of the pixels (``EopxManifest.canonical_payload`` covers
``image_sha3_512``). A figure drawn *into* the image and derived *from* the
image hash is a fixed point with no solution. Only pre-image fields are
admissible inputs, and the freeze is enforced by :data:`FIGURE_VERSION` inside
the domain-separation strings: changing the derivation means a new version, and
figures already printed keep verifying under the old one.

The single ML-DSA-87 signature already covers both ends: the inputs live in
``canonical_payload()``, and the rendering reaches it through
``image_sha3_512``. No new primitive, no new trust root.
"""

from __future__ import annotations

import hashlib
from typing import List, Sequence

from .metatron.field import hkdf_sha3_512

#: Derivation version. Baked into the domain strings — bump it and every
#: figure changes, which is why printed artifacts pin the version they were
#: minted under. FROZEN: v1 must keep producing v1 grids forever.
FIGURE_VERSION = 1

INFO_CONTENT = b"esoptron.figure.content.v1"
INFO_EPOCH = b"esoptron.figure.epoch.v1"

#: Grid geometry. 16 x 8 = 128 cells of 4 bits = 512 bits, i.e. exactly the
#: width of the SHA3-512 the KDF is built on — no counter mode, no truncation
#: bias, one output one grid.
GRID_W = 16
GRID_H = 8
CONTENT_ROWS = 6
EPOCH_ROWS = GRID_H - CONTENT_ROWS

#: Cell levels. One hex nibble per cell.
LEVELS = 16

#: Reference presentation, light -> dense. Normative *as a rendering*: two
#: implementations printing the same grid must agree. The digest is computed
#: over the levels (:func:`canonical_text`), never over these glyphs, so a
#: Unicode or block-element ramp is a free substitution.
ASCII_RAMP = " .`',:;!~+=*%#@$"

_FP_LEN = 32


def _check(name: str, value: bytes) -> bytes:
    if len(value) != _FP_LEN:
        raise ValueError(f"{name} must be {_FP_LEN} bytes, got {len(value)}")
    return value


def _nibbles(data: bytes) -> List[int]:
    out: List[int] = []
    for b in data:
        out.append(b >> 4)
        out.append(b & 0x0F)
    return out


def figure_grid(merkle_root: bytes, dilithium_pk_fp: bytes) -> List[List[int]]:
    """The artifact's cell grid: ``GRID_H`` rows of ``GRID_W`` levels.

    ``merkle_root`` and ``dilithium_pk_fp`` are the raw 32-byte values carried
    (hex-encoded) by the ``.eopx`` manifest. Rows ``0..CONTENT_ROWS-1`` are the
    content band; the remaining rows are the epoch band.
    """
    _check("merkle_root", merkle_root)
    _check("dilithium_pk_fp", dilithium_pk_fp)

    content_cells = CONTENT_ROWS * GRID_W
    epoch_cells = EPOCH_ROWS * GRID_W

    content = hkdf_sha3_512(ikm=merkle_root, salt=b"", info=INFO_CONTENT,
                            length=content_cells // 2)
    epoch = hkdf_sha3_512(ikm=merkle_root + dilithium_pk_fp, salt=b"",
                          info=INFO_EPOCH, length=epoch_cells // 2)

    cells = _nibbles(content) + _nibbles(epoch)
    return [cells[r * GRID_W:(r + 1) * GRID_W] for r in range(GRID_H)]


def canonical_text(grid: Sequence[Sequence[int]]) -> str:
    """The hashed form: one lowercase hex digit per cell, one line per row.

    Glyph-independent by construction — this, and not any rendering, is what
    :func:`figure_digest` covers.
    """
    if len(grid) != GRID_H or any(len(row) != GRID_W for row in grid):
        raise ValueError(f"grid must be {GRID_H}x{GRID_W}")
    if any(not 0 <= c < LEVELS for row in grid for c in row):
        raise ValueError(f"cell levels must be in [0, {LEVELS})")
    return "\n".join("".join(f"{c:x}" for c in row) for row in grid)


def figure_digest(grid: Sequence[Sequence[int]]) -> bytes:
    """SHA3-256 over :func:`canonical_text` — the figure's own fingerprint."""
    return hashlib.sha3_256(canonical_text(grid).encode("utf-8")).digest()


def figure_tag(grid: Sequence[Sequence[int]]) -> str:
    """Short human-comparable form of :func:`figure_digest` (8 hex chars).

    Small enough to print under a badge, large enough that two artifacts
    colliding by accident is not a practical concern. It is a *comparison*
    aid, not evidence: only recomputation from the `.eopx` decides.
    """
    return figure_digest(grid)[:4].hex()


def epoch_id(dilithium_pk_fp: bytes) -> str:
    """The issuing key's epoch tag (first 4 bytes of its fingerprint, hex).

    Two artifacts sharing an ``epoch_id`` were signed under the same key. The
    tag is a legible signal, never a proof of authority — that is the cosigned
    attestation chain's job (EPX-F §5).
    """
    return _check("dilithium_pk_fp", dilithium_pk_fp)[:4].hex()


def render_rows(grid: Sequence[Sequence[int]],
                ramp: str = ASCII_RAMP) -> List[str]:
    """Present ``grid`` with a glyph ramp (default :data:`ASCII_RAMP`).

    ``ramp`` must have exactly :data:`LEVELS` glyphs. Substituting a Unicode
    ramp changes what a reader sees and nothing else: the grid, the digest and
    the tag are untouched.
    """
    if len(ramp) != LEVELS:
        raise ValueError(f"ramp must have exactly {LEVELS} glyphs")
    return ["".join(ramp[c] for c in row) for row in grid]


__all__ = [
    "FIGURE_VERSION", "INFO_CONTENT", "INFO_EPOCH",
    "GRID_W", "GRID_H", "CONTENT_ROWS", "EPOCH_ROWS", "LEVELS", "ASCII_RAMP",
    "figure_grid", "canonical_text", "figure_digest", "figure_tag",
    "epoch_id", "render_rows",
]

File: pwa/src/lib/artifactFigure.ts
/**
 * EPX-F — TypeScript port of ``eopx.artifact_figure``.
 *
 * The artifact figure is a **cell grid**, not a picture: ``GRID_H`` rows of
 * ``GRID_W`` integer levels that any renderer may draw with any glyph set.
 * A phone recomputes it from a verified `.eopx` manifest and compares the tag
 * against what is printed on the object (EPX-F §7). Looking at the picture is
 * not a step.
 *
 * Two properties are load-bearing and must survive the port:
 *
 * * **Pre-image inputs only.** The grid derives from ``merkle_root`` and
 *   ``dilithium_pk_fp``, both fixed before the image exists. ``payload_hash``
 *   and ``image_sha3_512`` are downstream of the pixels, so a figure drawn
 *   into the image and derived from them is a fixed point with no solution
 *   (EPX-F §1).
 * * **Two bands.** Rows 0..5 depend on ``merkle_root`` alone, so an artifact
 *   keeps its face for life; rows 6..7 also take ``dilithium_pk_fp``, so a key
 *   rotation is legible without making the object a stranger (EPX-F §3.1).
 *
 * FROZEN at v1: ``FIGURE_VERSION`` is baked into the domain strings, and the
 * §9 vectors are normative. A failing vector is a regression or a deliberate
 * version bump — never an expected-value edit (EPX-F §8).
 */

import { concat, fromHex, hkdfSha3_512, sha3_256_, toHex } from "./crypto";

const utf8 = (s: string): Uint8Array => new TextEncoder().encode(s);

/** Derivation version, baked into the domain strings below. */
export const FIGURE_VERSION = 1;

export const INFO_CONTENT = utf8("esoptron.figure.content.v1");
export const INFO_EPOCH = utf8("esoptron.figure.epoch.v1");

/** Grid geometry. 16 × 8 = 128 cells of 4 bits = 512 bits, exactly one
 *  SHA3-512 width: no counter mode, no truncation bias, one output one grid. */
export const GRID_W = 16;
export const GRID_H = 8;
export const CONTENT_ROWS = 6;
export const EPOCH_ROWS = GRID_H - CONTENT_ROWS;

/** One hex nibble per cell. */
export const LEVELS = 16;

/** Reference presentation, light → dense. Normative *as a rendering*: two
 *  implementations printing the same grid with it must agree character for
 *  character. The digest covers the levels, never these glyphs. */
export const ASCII_RAMP = " .`',:;!~+=*%#@$";

const FP_LEN = 32;

function check(name: string, value: Uint8Array): Uint8Array {
  if (value.length !== FP_LEN) {
    throw new Error(`${name} must be ${FP_LEN} bytes, got ${value.length}`);
  }
  return value;
}

function nibbles(data: Uint8Array): number[] {
  const out: number[] = [];
  for (const b of data) {
    out.push(b >> 4);
    out.push(b & 0x0f);
  }
  return out;
}

/**
 * The artifact's cell grid: ``GRID_H`` rows of ``GRID_W`` levels.
 *
 * ``merkleRoot`` and ``dilithiumPkFp`` are the raw 32-byte values carried
 * hex-encoded by the `.eopx` manifest — decode them before calling.
 */
export function figureGrid(
  merkleRoot: Uint8Array,
  dilithiumPkFp: Uint8Array,
): number[][] {
  check("merkle_root", merkleRoot);
  check("dilithium_pk_fp", dilithiumPkFp);

  const contentCells = CONTENT_ROWS * GRID_W;
  const epochCells = EPOCH_ROWS * GRID_W;
  const empty = new Uint8Array(0);

  const content = hkdfSha3_512(merkleRoot, empty, INFO_CONTENT, contentCells / 2);
  const epoch = hkdfSha3_512(
    concat(merkleRoot, dilithiumPkFp),
    empty,
    INFO_EPOCH,
    epochCells / 2,
  );

  const cells = nibbles(content).concat(nibbles(epoch));
  const grid: number[][] = [];
  for (let r = 0; r < GRID_H; r += 1) {
    grid.push(cells.slice(r * GRID_W, (r + 1) * GRID_W));
  }
  return grid;
}

function assertGrid(grid: ReadonlyArray<ReadonlyArray<number>>): void {
  if (grid.length !== GRID_H || grid.some((row) => row.length !== GRID_W)) {
    throw new Error(`grid must be ${GRID_H}x${GRID_W}`);
  }
  for (const row of grid) {
    for (const c of row) {
      if (!Number.isInteger(c) || c < 0 || c >= LEVELS) {
        throw new Error(`cell levels must be integers in [0, ${LEVELS})`);
      }
    }
  }
}

/**
 * The hashed form: one lowercase hex digit per cell, rows joined by ``\n``,
 * no trailing newline. Glyph-independent by construction — this, and not any
 * rendering, is what {@link figureDigest} covers.
 */
export function canonicalText(
  grid: ReadonlyArray<ReadonlyArray<number>>,
): string {
  assertGrid(grid);
  return grid.map((row) => row.map((c) => c.toString(16)).join("")).join("\n");
}

/** SHA3-256 over {@link canonicalText} — the figure's own fingerprint. */
export function figureDigest(
  grid: ReadonlyArray<ReadonlyArray<number>>,
): Uint8Array {
  return sha3_256_(utf8(canonicalText(grid)));
}

/**
 * Short human-comparable form of {@link figureDigest} — 8 hex characters.
 *
 * A *comparison aid*, never evidence: only recomputation from the `.eopx`
 * decides. It is what a reader checks against the tag printed on the object.
 */
export function figureTag(
  grid: ReadonlyArray<ReadonlyArray<number>>,
): string {
  return toHex(figureDigest(grid).slice(0, 4));
}

/**
 * The issuing key's epoch tag — first 4 bytes of its fingerprint, hex.
 *
 * Two artifacts sharing an epoch id were signed under the same key. A legible
 * signal, never a proof of authority: that is the cosigned attestation chain's
 * job (EPX-F §5).
 */
export function epochId(dilithiumPkFp: Uint8Array): string {
  return toHex(check("dilithium_pk_fp", dilithiumPkFp).slice(0, 4));
}

/**
 * Present a grid with a glyph ramp (default {@link ASCII_RAMP}).
 *
 * ``ramp`` must have exactly {@link LEVELS} glyphs. Substituting a ramp
 * changes what a reader sees and nothing else: the grid, the digest and the
 * tag are untouched.
 *
 * The ramp is indexed by code point, not by UTF-16 unit, so a ramp containing
 * astral-plane glyphs still maps one glyph to one level.
 */
export function renderRows(
  grid: ReadonlyArray<ReadonlyArray<number>>,
  ramp: string = ASCII_RAMP,
): string[] {
  const glyphs = Array.from(ramp);
  if (glyphs.length !== LEVELS) {
    throw new Error(`ramp must have exactly ${LEVELS} glyphs`);
  }
  assertGrid(grid);
  return grid.map((row) => row.map((c) => glyphs[c]).join(""));
}

/**
 * Recompute the figure of a `.eopx` from its manifest fields, as hex strings.
 *
 * Convenience for the verifier path (EPX-F §7 steps 2–4): the manifest carries
 * these fields hex-encoded, and this is the shape a caller already holds after
 * a signature check.
 */
export function figureFromManifest(
  merkleRootHex: string,
  dilithiumPkFpHex: string,
): { grid: number[][]; tag: string; epoch: string } {
  const fp = fromHex(dilithiumPkFpHex.trim());
  const grid = figureGrid(fromHex(merkleRootHex.trim()), fp);
  return { grid, tag: figureTag(grid), epoch: epochId(fp) };
}
